### PR TITLE
Catch DKIM and DMARC filter exceptions and log them

### LINKED ISSUE #263: DKIMValidityFilter crashes
[afew 2.0.0]

I inserted DKIMValidityFilter into the chain. Afew unexpectedly crashed on a dry run of my mail. It looks like no DNS server responded which is odd because it responds from the command line:
```
$ dig -t txt dk1024-2012._domainkey.engage.vmware.com

; <<>> DiG 9.14.9 <<>> -t txt dk1024-2012._domainkey.engage.vmware.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 55598
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: ed80d888d7e6a6e4fc2276105e9487829dd3cee0b72f6c1b (good)
;; QUESTION SECTION:
;dk1024-2012._domainkey.engage.vmware.com. IN TXT

;; ANSWER SECTION:
dk1024-2012._domainkey.engage.vmware.com. 3600 IN TXT "v=DKIM1;" "p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC1TaNgLlSyQMNWVLNLvyY/neDgaL2oqQE8T5illKqCgDtFHc8eHVAU+nlcaGmrKmDMw9dbgiGk1ocgZ56NR4ycfUHwQhvQPMUZw0cveel/8EAGoi/UyPmqfcPibytH81NFtTMAxUeM4Op8A6iHkvAMj5qLf4YRNsTkKAVmJ5jEZwIDAQAB;"

;; Query time: 15 msec
;; SERVER: 198.82.247.66#53(198.82.247.66)
;; WHEN: Mon Apr 13 11:38:42 EDT 2020
;; MSG SIZE  rcvd: 338
```

Here is the traceback (on NixOS 19.09):
```
DEBUG:dkimpy:sig: {b'v': b'1', b'a': b'rsa-sha256', b'c': b'relaxed/relaxed', b's': b'dk1024-2012', b'd': b'engage.vmware.com', b'h': b'Message-ID:List-Unsubscribe:MIME-Version:From:To:Reply-To:Date:Subject:\r\n Content-Type', b'i': b'vmwareteam@engage.vmware.com', b'bh': b'0VxUCX2CqiU7uAwWiAZqWGHDvnYofDA+7ISD/j53iCA=', b'b': b'GIZ98rUkY37EwJ+3jPlsFwSFz0tgurUJH2VGzV2HVYqkfwxPX2eXa/IVxiJkwfks6nerGKqzkC0r\r\n   aBxMOcwyAAnPf+RTNKk7QmnX5kQlnwFjdrntj5KA8kxavb6ABcDzUzA66qBAZ7r9M+4hwo6N7R7s\r\n   q1mqkoYTJsxOqkC8gts='}
Traceback (most recent call last):
  File "/nix/store/7axc6gpph1fnjrwf02q6j4alh66m5n3i-afew-2.0.0/bin/.afew-wrapped", line 11, in <module>
    sys.exit(main())
  File "/nix/store/7axc6gpph1fnjrwf02q6j4alh66m5n3i-afew-2.0.0/lib/python3.7/site-packages/afew/commands.py", line 160, in main
    inner_main(args, database, query_string)
  File "/nix/store/7axc6gpph1fnjrwf02q6j4alh66m5n3i-afew-2.0.0/lib/python3.7/site-packages/afew/main.py", line 23, in main
    filter_.run(query_string)
  File "/nix/store/7axc6gpph1fnjrwf02q6j4alh66m5n3i-afew-2.0.0/lib/python3.7/site-packages/afew/filters/BaseFilter.py", line 60, in run
    self.handle_message(message)
  File "/nix/store/7axc6gpph1fnjrwf02q6j4alh66m5n3i-afew-2.0.0/lib/python3.7/site-packages/afew/filters/DKIMValidityFilter.py", line 38, in handle_message
    dkim_ok = all(map(verify_dkim, message.get_filenames()))
  File "/nix/store/7axc6gpph1fnjrwf02q6j4alh66m5n3i-afew-2.0.0/lib/python3.7/site-packages/afew/filters/DKIMValidityFilter.py", line 22, in verify_dkim
    return dkim.verify(message_bytes)
  File "/nix/store/pglvrcim2hqbf9l8fpvfab71wvpyjn7s-python3.7-dkimpy-0.9.2/lib/python3.7/site-packages/dkim/__init__.py", line 1266, in verify
    return d.verify(dnsfunc=dnsfunc)
  File "/nix/store/pglvrcim2hqbf9l8fpvfab71wvpyjn7s-python3.7-dkimpy-0.9.2/lib/python3.7/site-packages/dkim/__init__.py", line 869, in verify
    return self.verify_sig(sig, include_headers, sigheaders[idx], dnsfunc)
  File "/nix/store/pglvrcim2hqbf9l8fpvfab71wvpyjn7s-python3.7-dkimpy-0.9.2/lib/python3.7/site-packages/dkim/__init__.py", line 664, in verify_sig
    pk, self.keysize, ktag = load_pk_from_dns(name, dnsfunc)
  File "/nix/store/pglvrcim2hqbf9l8fpvfab71wvpyjn7s-python3.7-dkimpy-0.9.2/lib/python3.7/site-packages/dkim/__init__.py", line 423, in load_pk_from_dns
    s = dnsfunc(name)
  File "/nix/store/pglvrcim2hqbf9l8fpvfab71wvpyjn7s-python3.7-dkimpy-0.9.2/lib/python3.7/site-packages/dkim/dnsplug.py", line 85, in get_txt
    txt = _get_txt(unicode_name)
  File "/nix/store/pglvrcim2hqbf9l8fpvfab71wvpyjn7s-python3.7-dkimpy-0.9.2/lib/python3.7/site-packages/dkim/dnsplug.py", line 31, in get_txt_dnspython
    a = dns.resolver.query(name, dns.rdatatype.TXT,raise_on_no_answer=False)
  File "/nix/store/r40vvpd265ijvrzyac6pcq34qz68q185-python3.7-dnspython-1.16.0/lib/python3.7/site-packages/dns/resolver.py", line 1102, in query
    lifetime)
  File "/nix/store/r40vvpd265ijvrzyac6pcq34qz68q185-python3.7-dnspython-1.16.0/lib/python3.7/site-packages/dns/resolver.py", line 898, in query
    raise NoNameservers(request=request, errors=errors)
dns.resolver.NoNameservers: All nameservers failed to answer the query dk1024-2012._domainkey.engage.vmware.com. IN TXT: Server 198.82.247.66 UDP port 53 answered SERVFAIL; Server 198.82.247.34 UDP port 53 answered SERVFAIL
```

### PR DESCRIPTION
**Pull request by aidecoe**
Instead of interrupting afew tagging, just log errors.

Fixes #263.

### CODE REVIEW

**Comment by aidecoe** [comment]
Quick fix, but needs no more work like unit tests.

**Comment by GuillaumeSeren** [comment]
Well I tested the branch, here an example of the error on some message on the linux-kernel ML.

With master 22d035c
```
Traceback (most recent call last):
  File "/home/gseren/.local/bin/afew", line 11, in <module>
    load_entry_point('afew==3.0.1.dev2+g22d035c', 'console_scripts', 'afew')()
  File "/home/gseren/.local/lib64/python3.6/site-packages/afew-3.0.1.dev2+g22d035c-py3.6.egg/afew/commands.py", line 160, in main
    inner_main(args, database, query_string)
  File "/home/gseren/.local/lib64/python3.6/site-packages/afew-3.0.1.dev2+g22d035c-py3.6.egg/afew/main.py", line 19, in main
    filter_.run(query_string)
  File "/home/gseren/.local/lib64/python3.6/site-packages/afew-3.0.1.dev2+g22d035c-py3.6.egg/afew/filters/BaseFilter.py", line 55, in run
    self.handle_message(message)
  File "/home/gseren/.local/lib64/python3.6/site-packages/afew-3.0.1.dev2+g22d035c-py3.6.egg/afew/filters/DKIMValidityFilter.py", line 35, in handle_message
    dkim_ok = all(map(verify_dkim, message.get_filenames()))
  File "/home/gseren/.local/lib64/python3.6/site-packages/afew-3.0.1.dev2+g22d035c-py3.6.egg/afew/filters/DKIMValidityFilter.py", line 19, in verify_dkim
    return dkim.verify(message_bytes)
  File "/usr/lib64/python3.6/site-packages/dkim/__init__.py", line 1266, in verify
    return d.verify(dnsfunc=dnsfunc)
  File "/usr/lib64/python3.6/site-packages/dkim/__init__.py", line 869, in verify
    return self.verify_sig(sig, include_headers, sigheaders[idx], dnsfunc)
  File "/usr/lib64/python3.6/site-packages/dkim/__init__.py", line 664, in verify_sig
    pk, self.keysize, ktag = load_pk_from_dns(name, dnsfunc)
  File "/usr/lib64/python3.6/site-packages/dkim/__init__.py", line 423, in load_pk_from_dns
    s = dnsfunc(name)
  File "/usr/lib64/python3.6/site-packages/dkim/dnsplug.py", line 85, in get_txt
    txt = _get_txt(unicode_name)
  File "/usr/lib64/python3.6/site-packages/dkim/dnsplug.py", line 31, in get_txt_dnspython
    a = dns.resolver.query(name, dns.rdatatype.TXT,raise_on_no_answer=False)
  File "/usr/lib64/python3.6/site-packages/dns/resolver.py", line 1102, in query
    lifetime)
  File "/usr/lib64/python3.6/site-packages/dns/resolver.py", line 898, in query
    raise NoNameservers(request=request, errors=errors)
dns.resolver.NoNameservers: All nameservers failed to answer the query pfpt0818._domainkey.marvell.com. IN TXT: Server 127.0.0.53 UDP port 53 answered The DNS operation timed out.; Server 127.0.0.53 UDP port 53 answered The DNS operation timed out.; Server 127.0.0.53 UDP port 53 answered SERVFAIL
```

And I got no more error with your branch, but a lot of logs,
maybe it could be something the user can configure (fail on DKIM errors) ?

I will run it against a larger amount of mail, but the issue seem's fixed.

> Quick fix, needs no more work like unit tests.

Yes we have some tests now (as you see on the gentoo package),
so ideally it could be a very good oportunity to add some.

**Comment by GuillaumeSeren** [comment]
Maybe @flokli you want to have a look at this ?

**Comment by GuillaumeSeren** [comment]
Here the logged error with your branch
```
DEBUG:afew.filters.DKIMValidityFilter.DKIMValidityFilter:Adding tags dkim-fail to id:eb1f4ecbda4aa49a23bc459a37b8f160828f2919.foo@bar.com
DEBUG:dkimpy:sig: {b'v': b'1', b'a': b'rsa-sha256', b'c': b'relaxed/relaxed', b'd': b'bar.com', b'h': b'from : to : cc :\r\n subject : date : message-id : references : in-reply-to : content-type :\r\n content-id : content-transfer-encoding : mime-version', b's': b'pfpt0818', b'bh': b'MTgWmjDqBAQ16SXK7Xqo80CSSZHLL5odf6Rput0yktE=', b'b': b'vPRaAw+KFEzsJhQ8yBbxjqeRrDkU33+Yth+ZUMVgf5QSvLyw8xY0wMhSR5GXZx13GjB4\r\n 3SpxNnc8OucS4YyS6oZE9EJ+uZPDwYMlqPUAyWorpQJcsCH9RLOsePciT6ww6ByY7EJp\r\n YpK7/gdNWlJLoCayMtG1PDE4INKXG0Efh5SScQCuphvMkM4qTH5dkEXuVkn7GRQyvw3A\r\n Gru3u+dTeeOeqyUWk8p1Lpa0daWopeqOyvtdj5pi1QQrlIPoOUldlFsqbKj36UTJLyzh\r\n 0UCdiRrOh8GnwGvu6zeqoB56+llVoT0vtZNmGIIlV8p75+K43N+AtnzUPVg6T/JTz60e pg=='}
WARNING:afew.filters.DKIMValidityFilter.DKIMValidityFilter:Failed to verify DKIM of 'e4bed0c6be09443ef5f91ee0a084a4a558c3007foo@bar.com': The DNS operation timed out after 30.000890254974365 seconds (marked as 'dkim-fail')
```

**Comment by aidecoe** [comment]
Added tests.

**Comment by GuillaumeSeren** [comment]
Hey @aidecoe ,
thank you for adding the tests for this filter, this is a huge step forward,
and the goal of every filter tested is closer now.

I will review the diff.

File: afew/tests/test_dkimvalidityfilter.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
+"""Test suite for DKIMValidityFilter.
```
Comment (GuillaumeSeren) [inline afew/tests/test_dkimvalidityfilter.py:1]: I think we need the license template here:
```
# SPDX-License-Identifier: ISC
```

File: afew/tests/test_dkimvalidityfilter.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+from unittest import mock
+
+import dkim
+import dns.exception
+
+from afew.Database import Database
+from afew.filters.DKIMValidityFilter import DKIMValidityFilter
+
+
+class _AddTags:  # pylint: disable=too-few-public-methods
```
Comment (GuillaumeSeren) [inline afew/tests/test_dkimvalidityfilter.py:14]: I am wondering if we should centralize that sort of config in a pylintrc, but it is a lesser detail.

File: afew/tests/test_dkimvalidityfilter.py (hunk 1 above)
Comment (aidecoe) [reply afew/tests/test_dkimvalidityfilter.py:1]: Do we really need to do that? Cannot we just have top level LICENCE file, please?

File: afew/tests/test_dkimvalidityfilter.py (hunk 2 above)
Comment (aidecoe) [reply afew/tests/test_dkimvalidityfilter.py:14]: well, that depends whether this is exception of the rule or actually "too-few-public-methods" warning isn't something we'd like to bothered at all. I'd say this warning is absolute non-sense.

**Comment by aidecoe** [comment]
> I have tested the branch filter on the kernel ML,
> beside some dns timeout this works fine.
> I also tested the test suite.

Could you paste the traceback? That should also be caught.

File: afew/tests/test_dkimvalidityfilter.py (hunk 1 above)
Comment (GuillaumeSeren) [reply afew/tests/test_dkimvalidityfilter.py:1]: Yes sure

**Comment by GuillaumeSeren** [comment]
Here some logs:

```
$ afew --tag -vv -- 'tag:lists/linux-kernel and date:2020-04-17..2020-04-17'
```
DNS operation timed out
```
DEBUG:afew.filters.DKIMValidityFilter.DKIMValidityFilter:Adding tags dkim-ok to id:CAKMK7uHXtcMFv6LpCEjTg5V-TXbkt2eLyitZoCfSc-nMnwJbuA@mail.gmail.com                               
DEBUG:dkimpy:sig: {b'v': b'1', b'a': b'rsa-sha256', b'c': b'relaxed/relaxed', b'd': b'nifty.com', b's': b'dec2015msa', b't': b'1587147200', b'bh': b'xO9EX2FjdnmulH032XwMxEkWt0+Xozs
HmF+IUr7+mzc=', b'h': b'References:In-Reply-To:From:Date:Subject:To:Cc:From', b'b': b'iTEENjHNsaXV5k8y6Hsan5NYkMb5BI0tt0uPGXby6Cq+LA/B5iSRxU5ibZleFxd6b\r\n         Sbj9e8pJ5ggzBhJc
x4A6dGz8RKU6ExSLCT5o74nIRaPEGKpQR9RaoHxIgogTTDhdD0\r\n         aFGcg3HKftbbb+WwTUORphdcrSA1it7V/+FTWNqEgl8fnUZ0i9yaEOxA2R6GL/oL3D\r\n         0b048c4r4xtHzxyGo/emxB+wM8hzJORGAhV4Jn
bFijTIXTT0v6QOmcvnjSV0CoKs22\r\n         u+gjOjeica0StuIbzA7PfJnOmmuPmwwO8GC7iXunK9AEqqDVsxgABz6GjI3rHKzI83\r\n         Lfs2AxPtuJ65Q=='}                                           
WARNING:afew.filters.DKIMValidityFilter.DKIMValidityFilter:Failed to verify DKIM of 'CAK7LNAQYxtcNinV7JR+c8Pn9Rp1g0TxJ7m_mOFNOJQsB=OiAoA@mail.gmail.com': The DNS operation timed out after 5.000402212142944 seconds (marked as 'dkim-fail') 
```
dkim-fail
```
DEBUG:dkimpy:sig: {b'v': b'1', b'a': b'rsa-sha256', b'c': b'relaxed/relaxed', b'd': b'crapouillou.net', b's': b'mail', b't': b'1587142846', b'h': b'from:from:sender:reply-to:subjec
t:subject:date:date:\r\n         message-id:message-id:to:to:cc:cc:mime-version:mime-version:\r\n         content-type:content-transfer-encoding:content-transfer-encoding:\r\n     
    in-reply-to:in-reply-to:references:references', b'bh': b'8jnFrNHMJwYTRfGU+bxklpH8fY874eOvwZEWt6NCZZM=', b'b': b'xYaAzoSuObd/ydytGqBUBmjqrbvJAXHDUZvE8cA7wGGfTu3m82DOx4XdnQ3jfS8v
54wuj8\r\n        GGhtJpSwyJWSVo3nbSwy0MBQM2Fj9FJgxPqmduDc7GO3tSZsT4PFyA/5cw/C+okjws+xom\r\n        bMnrpv6IkoNPpqO1e9CC4YoxFe3Xp8g='}                                              
DEBUG:dkimpy:bh: b'8jnFrNHMJwYTRfGU+bxklpH8fY874eOvwZEWt6NCZZM='                                                                                                                    
DEBUG:dkimpy:b'DKIM-Signature' valid: False                                                                                                                                         
DEBUG:afew.filters.DKIMValidityFilter.DKIMValidityFilter:Adding tags dkim-fail to id:20200417170040.174319-3-paul@crapouillou.net 
```
DKIM valid False
```
DEBUG:afew.filters.DKIMValidityFilter.DKIMValidityFilter:Adding tags dkim-ok to id:20200417154738.GT9767@localhost.localdomain                                                      
DEBUG:dkimpy:sig: {b'v': b'1', b'a': b'rsa-sha256', b'q': b'dns/txt', b'c': b'relaxed/relaxed', b'd': b'infradead.org', b's': b'bombadil.20170209', b'h': b'In-Reply-To:Content-Type
:MIME-Version\r\n        :References:Message-ID:Subject:Cc:To:From:Date:Sender:Reply-To:\r\n        Content-Transfer-Encoding:Content-ID:Content-Description', b'bh': b'HDRDbJd2M4qi
CgjpoZQhZ31JvaoOb1sTaN9fveDaTJE=', b'b': b'Ysjg7rsRv8f4AtxBOZkNUdzkap\r\n        jjG9O7xTEECfR1lzCTGcK40z/woI/GZqIa0YJhR0qkMw+62vqvxwSLJ8qx/y5gpu9BOwINgi5BT8B\r\n        5wBi/VqyuC
Zq/RAdNMPahC1amqdi3pQHOobCM9s2Td7hFQ1YdffHNrq777ocJJa45J8iVG6z0bIAD\r\n        VzgH+xK0aPb1LT3lZbrpRicKuPv28WciTB9BNqH9yxQ4fFiLTeVxVQCdcMnri6yCQR0B5SAPg0Ojc\r\n        v+m7WUhL6pjR
Mf9jfv3Ae77lLFUY+FnOZUBl5XvZLH/ZNTedT5aUtkf1DZTzjG3eOloV4r6IioGfi\r\n        1Ol16sPg=='}                                                                                           
DEBUG:dkimpy:bh: b'HDRDbJd2M4qiCgjpoZQhZ31JvaoOb1sTaN9fveDaTJE='                                                                                                                    
DEBUG:dkimpy:b'DKIM-Signature' valid: False  
```

**Comment by GuillaumeSeren** [comment]
@aidecoe 
What do think to become a co-maintainer on the project ?

You help a lot and it make sense to me.

File: afew/tests/test_dkimvalidityfilter.py (hunk 1 above)
Comment (flokli) [reply afew/tests/test_dkimvalidityfilter.py:1]: We don't need the SPDX License Identifier everywhere, and I personally wouldn't add it to new files.

We replaced the verbose license texts in each file with SPDX identifiers - as a compromise, given we can't easily remove other authors' copyright headers/licensing text, but SPDX headers are considered equivalent by most.

File: afew/tests/test_dkimvalidityfilter.py (hunk 2 above)
Comment (flokli) [reply afew/tests/test_dkimvalidityfilter.py:14]: In some cases this makes sense, for example to detect classes with a single method which should better be a function.

IMHO, it's perfectly valid to add this for something as synthetic as test tooling.

**Comment by aidecoe** [comment]
Those logs look valid. You don't actually experience uncaught exception. You only have warning logs with exception message which explains reason why it's marked as 'dkim-failed'. It's important it's logged as a warning because the verification fails not due to signature being invalid, but due to some other reasons.

Regards becoming co-maintainer - I'll think about it. If I'd continue to make more contributions that would make more sense.

**Comment by GuillaumeSeren** [comment]
Hey

> Regards becoming co-maintainer - I'll think about it. If I'd continue to make more contributions that would make more sense.

Sure thank you for this great contribution,
I sent you an invite you decide when or if you want to accept it :)

**Pull request merged**